File: axon-frontend/src/types/music.ts
/** Music mode types (YouTube-backed, provider-agnostic shape). */

export interface MusicTrack {
  videoId: string;
  title: string;
  artist: string;
  thumbnailUrl: string | null;
  durationSec: number;
}

export type RepeatMode = "off" | "one" | "all";

export interface MusicState {
  current: MusicTrack | null;
  queue: MusicTrack[];
  isPlaying: boolean;
  volume: number;
  shuffle: boolean;
  repeatMode: RepeatMode;
  positionSec: number;
  muted: boolean;
}

export interface MusicHistoryEntry {
  id: string;
  videoId: string;
  title: string;
  artist: string;
  thumbnailUrl: string | null;
  durationSec: number;
  playedAt: string;
}

export interface MusicPlaylist {
  id: string;
  name: string;
  playlistType: string;
  tracks: MusicTrack[];
}

export function normalizeTrack(raw: Record<string, unknown>): MusicTrack {
  return {
    videoId: String(raw.videoId ?? raw.video_id ?? ""),
    title: String(raw.title ?? "Unknown"),
    artist: String(raw.artist ?? ""),
    thumbnailUrl: (raw.thumbnailUrl ?? raw.thumbnail_url ?? null) as string | null,
    durationSec: Number(raw.durationSec ?? raw.duration_sec ?? 0),
  };
}

export function normalizeMusicState(raw: Record<string, unknown>): MusicState {
  const currentRaw = raw.current as Record<string, unknown> | null | undefined;
  const queueRaw = (raw.queue as Record<string, unknown>[] | undefined) ?? [];
  return {
    current: currentRaw ? normalizeTrack(currentRaw) : null,
    queue: queueRaw.map(normalizeTrack),
    isPlaying: Boolean(raw.isPlaying ?? raw.is_playing),
    volume: Number(raw.volume ?? 70),
    shuffle: Boolean(raw.shuffle),
    repeatMode: (raw.repeatMode ?? raw.repeat_mode ?? "off") as RepeatMode,
    positionSec: Number(raw.positionSec ?? raw.position_sec ?? 0),
    muted: Boolean(raw.muted),
  };
}

export function formatDuration(seconds: number): string {
  const s = Math.max(0, Math.floor(seconds));
  const m = Math.floor(s / 60);
  const r = s % 60;
  return `${m}:${String(r).padStart(2, "0")}`;
}

File: axon-backend/app/schemas/music.py
"""Music API schemas."""

from __future__ import annotations

from typing import Literal

from pydantic import Field

from app.schemas.common import CamelModel

RepeatMode = Literal["off", "one", "all"]
PlaylistType = Literal["favorites", "recent", "workout", "relax", "study", "custom"]


class MusicTrack(CamelModel):
    video_id: str
    title: str
    artist: str = ""
    thumbnail_url: str | None = None
    duration_sec: int = 0


class MusicSearchResult(CamelModel):
    tracks: list[MusicTrack]
    query: str


class MusicPlayRequest(CamelModel):
    query: str | None = None
    video_id: str | None = None
    track: MusicTrack | None = None
    autoplay_queue: bool = True


class MusicVolumeRequest(CamelModel):
    volume: int | None = Field(default=None, ge=0, le=100)
    delta: int | None = None
    mute: bool | None = None


class MusicSeekRequest(CamelModel):
    position_sec: float = Field(..., ge=0)


class MusicQueueUpdateRequest(CamelModel):
    queue: list[MusicTrack]


class MusicStateResponse(CamelModel):
    current: MusicTrack | None = None
    queue: list[MusicTrack] = Field(default_factory=list)
    is_playing: bool = False
    volume: int = 70
    shuffle: bool = False
    repeat_mode: RepeatMode = "off"
    position_sec: float = 0
    muted: bool = False


class MusicHistoryEntry(CamelModel):
    id: str
    video_id: str
    title: str
    artist: str = ""
    thumbnail_url: str | None = None
    duration_sec: int = 0
    played_at: str


class MusicHistoryResponse(CamelModel):
    items: list[MusicHistoryEntry]
    total: int


class MusicPlaylist(CamelModel):
    id: str
    name: str
    playlist_type: PlaylistType
    tracks: list[MusicTrack] = Field(default_factory=list)


class MusicPlaylistsResponse(CamelModel):
    playlists: list[MusicPlaylist]
